### DISCUSSION TITLE
Thoughts on using a git repo as a source for links to archive?

### DISCUSSION THREAD

**Discussion by infogulch** [ArchiveBox/ArchiveBox#1010] [Q&A] [answered]
I would like to watch a git branch as a source of links to extract (from markdown, source, plain text, etc., files) and archive.

From what I can tell this is not a functionality that ArchiveBox supports natively, and I'll need to develop some of the solution myself. I guess this would require some kind of 'ci'-like automation that extract links and pipes them to `archivebox add`.

Any ideas/opinions on how should I approach this?

**Comment by pirate** [comment, marked as answer]
If you just need links at the top level or within a hop of the README, do something like this:
 `archivebox schedule --every=day --depth=1 'https://github.com/path/to/repo`
 or
 `archivebox schedule --every=day 'https://raw.githubusercontent.com/path/to/specific/file/README.md`
 
 If you need to recurse fully down the directory tree and scan every single file then add something to cron like this:

`cd /path/to/git/dir; git pull; cat **/** | archivebox add`

**Reply by infogulch** [reply]
I think I want something closer to "recurse fully down the directory tree and scan every single file".

So I can either set up the cron job, or otherwise make sure that all files with links are eventually reachable from README, and set `--depth=[very high]`.

Ideally I'd like to have it scan the repo very frequently, asymptotically approaching live. Would this rescan the whole repo every time? Should I be concerned about how long it takes if there are thousands of links? If it were very highly integrated I suppose it could filter to only rescan files that have changed since the last scanned commit, I guess I could write something that would do that with the cron job.

Edit: Maybe like `git pull; git diff -z --name-only HEAD HEAD~1 | xargs -0 cat | archivebox add`

**Reply by pirate** [reply]
It will automatically skip links that were previously archived, no need to filter those by hand. Also `--depth=1` is only implemented up to 1, if you want to recurse futher you need to build the crawler yourself and pipe the URLs into archivebox.

**Reply by infogulch** [reply]
Thanks for adding details. I think this is my favorite solution so far:

`git pull -f; git diff -z --name-only master@{1}..master | xargs -0 cat | archivebox add`

1. Pull the latest changes
2. List out the names of files that were changed between the previous pull and this one
   * This works even if it pulled multiple commits (master@{1})
   * Should also be safe to weird paths (-z / -0)
4. Cat the current contents of all the changed files into archivebox add

This is nice because it scales better; in general it should only send O(changes) data to archivebox instead of O(repo-size) every time cron is triggered (which I might make as frequent as every 5 minutes).

This will be pure automation for me so I'd prefer to limit the impact of the repo size as much as possible. Future enhancements in this direction could be: 1. Make it work from a bare repo (git cat-file?), 2. Use shallow clones and [prune old commits](https://stackoverflow.com/a/46004595).

**Reply by pirate** [reply]
Yeah thats a good solution 👍
